### PR TITLE
fabric: Change how memory registration support is exposed

### PR DESCRIPTION
**Pull request by shefty**
Memory registration is currently exposed using 5 MR bits, including
caps, flags, and mode bits: FI_MR_OFFSET, FI_MR_KEY, FI_DYNAMIC_MR,
FI_LOCAL_MR, and FI_PROV_MR_ATTR.  Because of the combinations, this
can result in branches in the provider in data transfer operations.
For example, the provider needs to check if a key corresponds to
a virtual address or an offset based address.  It may also need to
check if the memory is pinned or might reference a dynamic MR.

To simplify the usage for both the application and the provider,
we replace all MR bits with a new domain attribute field.  The
new field is an enum that conveys how memory registration works
for the provider at a broad level.

Signed-off-by: Sean Hefty <sean.hefty@intel.com>

### CODE REVIEW

**Comment by shefty** [comment]
@goodell @j-xiong @jithinjosepkl @a-ilango -- This change affects the API and all providers.  It changes (simplifies) how memory registration is reported to apps and allows greater optimization in the provider.  Please review the proposed changes to the API and providers.  I will be updating the test apps to use this proposal.  I felt that this change was too high of a risk for rc5, but would like to create an rc6 later this week with this change, before freezing 1.0.

For the proposal, I selected only 2 memory registration modes.  More modes can be added, if needed.  But I'd like to keep the number small.  These changes were, in part, identified as part of the process of cleaning up the flag space and designing internal helper modules.

File: man/fi_domain.3.md [hunk 1]
```diff
@@ -390,6 +391,29 @@ either address vector format.  In this case, a provider may return
...
+: Only scalable memory registration operations are requested or supported.
+
+Buffers used in data transfer operations may require notifying the provider
+of their use before a data transfer can occur.  The mr_mode field indicates
+the type of memory registration that is required, and when registration is
+necessary.  Applications that require the use of a specific registration mode
+should set the domain attribute mr_mode to the necessary value when calling
+fi_getinfo.  The value FI_MR_UNSPEC may be used to indicate support for any
+registration mode.
+
```
Comment (jsquyres) [inline man/fi_domain.3.md]: Would probably be useful to put some kind of "See fi_mr(3) for more discussion of the basic and scalable modes," or somesuch.

File: man/fi_mr.3.md [hunk 2]
```diff
@@ -98,45 +98,49 @@ granted for access by fabric resources.  A memory buffer must be
...
+
+  Memory regions registered as the target of RMA and atomic operations are
+  associated with a MR key selected by the application.  The resulting memory
+  region will be accessible by remote peers starting at a base address of 0.
+  Because scalable registration mode refers to memory regions, versus data
+  buffers, the address ranges given for a registration request do not need
+  to map to data buffers allocated by the application at the time the
+  registration call is made.  That is, an application can register any
+  range of addresses in their virtual address space, whether or not those
+  addresses are back by physical pages or have been allocated.
```
Comment (jsquyres) [inline man/fi_mr.3.md]: ... back***ed*** by physical pages...

**Comment by shefty** [comment]
I know I reduced the PSM provider to scalable support only.  (I was figuring why would an app want basic if they can have scalable.  Btw, I'm not sold on these names.)   I will modify the changes to have psm support both, plus add the missing domain_attr.

**Comment by j-xiong** [comment]
That will be great. The basic mode support is useful because otherwise an application coded to basic mode only won't work with the provider.

**Comment by hppritcha** [comment]
I think this is going in the right direction, but there's a significant case missing.  Namely, the case when a provider does not need to have memory registered for messaging (send/recv), but does for RMA and atomic ops.  Before, we had FI_LOCAL_MR mode bit, which if a provider didn't need, could be cleared on output from ```fi_getinfo```.   With the proposal in its current state, that flexibility is lost both to the provider and to the application.  

I'd like to have a FI_MR_RMA attribute wherein the application only needs to explicitly register memory for one-sided operations.  
 
I think this PR warrants discussion at one of the weeklies prior to merging in to master.  Right now the options seem to reflect either having only a simple I/O MMU implementation for a device, or a device which is a caching agent within the caching domain of the host processor(s).  I think it would be better for the API to reflect more of a continuum of I/O MMU capabilities.  The previous mode bit method, although more complicated, did provide more flexibility.

**Comment by jithinjosepkl** [comment]
Thanks Sean, Sockets provider changes look good.

**Comment by a-ilango** [comment]
Verbs provider changes look fine.

**Comment by shefty** [comment]
The main issue with the MR bits is that it introduces branches into the code in multiple places.  In practice, I don't foresee apps mixing registration mode, where some registrations use the virtual address, others use offsets, etc.  We can definitely add more enum values, or maybe find a way for the app to indicate its options up front using MR bits.  The two cases that I defined were based on what it appeared to me the current providers could do.  We should definitely discuss this in tomorrow's ofiwg.  But to get things started, these are the MR flags...  :)

FI_MR_OFFSET - I think the provider needs to know this when the domain or endpoint is created.  It affects how the provider locates the application buffer on all incoming RMA/atomic operations.

FI_MR_KEY - Similar thoughts as using OFFSET.  One of the issues is if the framework ever tries to support both these options over providers that do not natively support them (e.g. verbs), then the domains must maintain some sort of remote MR mapping table.

FI_DYNAMIC_MR - Eh, it seems like either the provider supports this or it doesn't.  This option would be difficult for the framework to support without provider support without a significant loss of performance.

FI_LOCAL_MR - This was a mode, so the provider knows up front what behavior is expected.

FI_PROV_MR_ATTR - This was also a mode.

Based on this, maybe the only change that's needed from my proposal is to put back the FI_LOCAL_MR mode bit?

File: man/fi_domain.3.md (hunk 1 above)
Comment (shefty) [inline man/fi_domain.3.md]: There's a reference at the top that directs to fi_mr(3).  Were you wanting a separate comment later?  Or just an adjustment to that comment?

**Comment by hppritcha** [comment]
Putting the FI_LOCAL_MR bit back in would address the concerns I have.

**Comment by shefty** [comment]
I updated the PR to address the concerns with the PSM provider and to return the FI_LOCAL_MR mode bit.  I still removed the FI_PROV_MR_ATTR mode bit, because it was defined with a weird duality (an app could force it to be enabled).  The FI_MR_OFFSET and FI_MR_KEY registration flags and FI_DYNAMIC_MR cap are now assumed as part of the mr_mode attribute, with only 2 of the 8 combinations defined.

This should still make things easier on both the apps and providers

**Comment by j-xiong** [comment]
Now the psm changes look fine to me.

File: man/fi_domain.3.md [hunk 3]
```diff
@@ -390,6 +391,29 @@ either address vector format.  In this case, a provider may return
 FI_AV_UNSPEC to indicate that either format is supportable, or may return
 another AV type to indicate the optimal AV type supported by this domain. 
 
+## Memory Registration Mode (mr_type)
```
Comment (goodell) [inline man/fi_domain.3.md]: I think `mr_type` should be `mr_mode`.

Comment (shefty) [inline man/fi_domain.3.md]: Thanks - updated

**Comment by goodell** [comment]
Seems OK to me.  I can update the usnic provider once this lands in master.

**Comment by shefty** [comment]
I've tested these changes for the sockets and verbs providers with an updated version of fabtests.

**Pull request merged**